### PR TITLE
Add Azure AMQP/MQTT/HTTP examples + fix IBM Quickstart

### PR DESCRIPTION
**Pull request by ThomasLyetPro**
Signed-off-by: Thomas Lyet <thomas.lyet@intel.com>

### CODE REVIEW

**Comment by ThomasLyetPro** [comment]
Please don't accept the pull request until review is done internally

File: bluemix-quickstart/cpp/main.cpp [hunk 1]
```diff
@@ -122,19 +125,17 @@ void delivery_complete(void * context, MQTTClient_deliveryToken dt)
  * @param cause The reason for the disconnection.
  * Currently, <i>cause</i> is always set to NULL.
  */
-void connection_lost(void * context, char* cause)
-{
+void connection_lost(void * context, char* cause) {
 	printf("Connection lost\n");
-	exit(MQTTCLIENT_DISCONNECTED);
+	exit (MQTTCLIENT_DISCONNECTED);
```
Comment (marc2905) [inline bluemix-quickstart/cpp/main.cpp]: remove that space

Comment (marc2905) [inline bluemix-quickstart/cpp/main.cpp]: remove that space

File: azure-http/sample.json [hunk 2]
```diff
@@ -0,0 +1,24 @@
+{
+   "name":"IBM* Bluemix Quickstart",
```
Comment (marc2905) [inline azure-http/sample.json]: Bluemix?

File: azure-http/sample.json [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+      ],
+      "libs":[
+         "mraa",
+         "upm",
+         "iothub_client"
+      ]
+   },
+   "additional_fields":[
+      {
+         "id":"connectionString",
```
Comment (marc2905) [inline azure-http/sample.json]: Better description of the fields

File: azure-http/sample.json [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+   "description":"This project template allow you to test the Microsoft* Azure* IoT Hub.\nIt will read the temperature from the Grove Temperature Sensor then send it to the Azure* IoT Hub.\n\nFirst steps:\nCreate an IoT HuB following the instruction in https://github.com/Azure/azure-iot-sdks/blob/master/doc/setup_iothub.md. \nCreate the a connection string for a device following the instruction in https://github.com/Azure/azure-iot-sdks/blob/master/doc/manage_iot_hub.md.\nRun the example and check the results inside your Azure* IoT Hub\n\n* Other names and brands may be claimed as the property of others.",
+   "date":"2016-11-04",
+   "platform":{
+      "hwreq":[
+         "Grove Temperature Sensor (GroveTemp) connected to the Grove Base Shield Port A0"
+      ],
+      "libs":[
+         "mraa",
+         "upm",
+         "iothub_client"
```
Comment (marc2905) [inline azure-http/sample.json]: Are these all libs, nothing from the SDK?

File: azure-http/cpp/main.cpp [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+			ENUM_TO_STRING(IOTHUB_CLIENT_CONFIRMATION_RESULT, result));
+
+	/* Some device specific action code goes here... */
+	callbackCounter++;
+	IoTHubMessage_Destroy(eventInstance->messageHandle);
+}
+
+/**
+ * This project template allow you to test the Microsoft* Azure* IoT Hub.
+ * It will read the temperature from the Grove Temperature Sensor then send
```
Comment (marc2905) [inline azure-http/cpp/main.cpp]: It will read the temperature from the Grove Temperature Sensor **and** send...

File: azure-http/cpp/main.cpp [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+ * it to the Azure* IoT Hub.
+ *
+ * First steps:
+ * - Create an IoT HuB following the instruction in
+ * https://github.com/Azure/azure-iot-sdks/blob/master/doc/setup_iothub.md.
+ * - Create the a connection string for a device following the instruction in
```
Comment (marc2905) [inline azure-http/cpp/main.cpp]: Create a connection string

File: azure-http/cpp/main.cpp [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+			&& (platform != mraa::INTEL_EDISON_FAB_C)) {
+		std::cerr << "ERROR: Unsupported platform, exiting" << std::endl;
+		return mraa::ERROR_INVALID_PLATFORM;
+	}
+
+	// temperature sensor connected to A0 (analog in)
+	upm::GroveTemp* temp_sensor = new upm::GroveTemp(0);
+	if (temp_sensor == NULL) {
+		fprintf(stderr,
+				"ERROR: MRAA couldn't initialize Analog interface, exiting");
```
Comment (marc2905) [inline azure-http/cpp/main.cpp]: Check indentation

File: azure-http/cpp/main.cpp [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+	}
+
+	IOTHUB_CLIENT_LL_HANDLE iotHubClientHandle;
+
```
Comment (marc2905) [inline azure-http/cpp/main.cpp]: do you need so much space between the variables?

File: bluemix-quickstart/cpp/main.cpp [hunk 9]
```diff
@@ -31,20 +31,22 @@ extern "C" {
...
-const char * clientID =
-		((std::string ("d:quickstart:iotquick-edison:")) + DEVICE_ID).c_str();
+const char * clientID = strcat(
+		strcpy(new char[50], "d:quickstart:iotquick-edison:"), DEVICE_ID);
 // To authenticate with a token, set the username to "use-token-auth" and the
 // password to the token which you get when you register your device in IBM Bluemix.
 const char * username = NULL;
 const char * password = NULL;
-const char * host = "tcp://quickstart.messaging.internetofthings.ibmcloud.com:1883";
+const char * host =
```
Comment (marc2905) [inline bluemix-quickstart/cpp/main.cpp]: why did you move that to the next line?

File: azure-http/sample.json [hunk 10]
```diff
@@ -0,0 +1,24 @@
+{
+   "name":"IBM* Bluemix Quickstart",
+   "category":"Cloud",
+   "tag":"cloud",
+   "description":"This project template allow you to test the Microsoft* Azure* IoT Hub.\nIt will read the temperature from the Grove Temperature Sensor then send it to the Azure* IoT Hub.\n\nFirst steps:\nCreate an IoT HuB following the instruction in https://github.com/Azure/azure-iot-sdks/blob/master/doc/setup_iothub.md. \nCreate the a connection string for a device following the instruction in https://github.com/Azure/azure-iot-sdks/blob/master/doc/manage_iot_hub.md.\nRun the example and check the results inside your Azure* IoT Hub\n\n* Other names and brands may be claimed as the property of others.",
+   "date":"2016-11-04",
+   "platform":{
+      "hwreq":[
+         "Grove Temperature Sensor (GroveTemp) connected to the Grove Base Shield Port A0"
```
Comment (marc2905) [inline azure-http/sample.json]: GroveLED

File: azure-http/cpp/main.cpp [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+				if (Map_AddOrUpdate(propMap, "PropName", propText) != MAP_OK) {
+					fprintf(stderr, "WARN: Map_AddOrUpdate Failed!\n");
+				}
+
+				if (IoTHubClient_LL_SendEventAsync(iotHubClientHandle,
+						messages[iterator].messageHandle,
+						SendConfirmationCallback, &messages[iterator])
+						!= IOTHUB_CLIENT_OK) {
+					fprintf(stderr,
+							"ERROR: IoTHubClient_LL_SendEventAsync failed to send message [%zu]! to IoT Hub\n",
```
Comment (marc2905) [inline azure-http/cpp/main.cpp]: put the message in two lines

File: azure-http/cpp/main.cpp [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+							iterator);
+				} else {
+					printf(
+							"IoTHubClient_LL_SendEventAsync accepted message [%zu] for transmission to IoT Hub.\n",
+							iterator);
```
Comment (marc2905) [inline azure-http/cpp/main.cpp]: put the message in two lines

File: azure-http/cpp/main.cpp [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+			}
+		}
+		IoTHubClient_LL_DoWork(iotHubClientHandle);
+		ThreadAPI_Sleep(1);
+
+		iterator++;
+	} while (g_continueRunning);
+
+	printf(
+			"iothub_client_sample_http has gotten quit message, call DoWork %d more time to complete final sending...\n",
```
Comment (marc2905) [inline azure-http/cpp/main.cpp]: put the message in two lines

File: azure-http/cpp/main.cpp [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+	mraa::Platform platform = mraa::getPlatformType();
+	if ((platform != mraa::INTEL_GALILEO_GEN1)
+			&& (platform != mraa::INTEL_GALILEO_GEN2)
+			&& (platform != mraa::INTEL_EDISON_FAB_C)) {
+		std::cerr << "ERROR: Unsupported platform, exiting" << std::endl;
+		return mraa::ERROR_INVALID_PLATFORM;
+	}
+
+	// temperature sensor connected to A0 (analog in)
+	upm::GroveTemp* temp_sensor = new upm::GroveTemp(0);
```
Comment (marc2905) [inline azure-http/cpp/main.cpp]: init of the GroveLED missing

File: azure-http/cpp/main.cpp [hunk 15]
```diff
@@ -0,0 +1,257 @@
...
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <iostream>
+
+/* This sample uses the _LL APIs of iothub_client for example purposes.
+ That does not mean that HTTP only works with the _LL APIs.
+ Simply changing the using the convenience layer (functions not having _LL)
```
Comment (marc2905) [inline azure-http/cpp/main.cpp]: the using the

File: azure-http/cpp/main.cpp [hunk 16]
```diff
@@ -0,0 +1,257 @@
...
+ and removing calls to _DoWork will yield the same results. */
+
```
Comment (marc2905) [inline azure-http/cpp/main.cpp]: Describe what is doing

File: azure-http/cpp/main.cpp [hunk 17]
```diff
@@ -0,0 +1,257 @@
...
+	}
+
+	// Retrieve properties from the message
+	mapProperties = IoTHubMessage_Properties(message);
+	if (mapProperties != NULL) {
+		const char* const * keys;
+		const char* const * values;
+		size_t propertyCount = 0;
+		if (Map_GetInternals(mapProperties, &keys, &values, &propertyCount)
+				== MAP_OK) {
```
Comment (marc2905) [inline azure-http/cpp/main.cpp]: put in one line

File: azure-http/cpp/main.cpp [hunk 18]
```diff
@@ -0,0 +1,257 @@
...
+} EVENT_INSTANCE;
+
+/**
+ * Message call back function that will be called when an answer from
+ * the Microsoft* Azure* IoT Hub is received.
+ * The received data will then be displayed in the standard output.
+ */
+static IOTHUBMESSAGE_DISPOSITION_RESULT ReceiveMessageCallback(
+		IOTHUB_MESSAGE_HANDLE message, void* userContextCallback) {
+	int* counter = (int*) userContextCallback;
```
Comment (marc2905) [inline azure-http/cpp/main.cpp]: pointer null check?

File: bluemix-quickstart/cpp/main.cpp [hunk 19]
```diff
@@ -31,20 +31,22 @@ extern "C" {
...
+		"tcp://quickstart.messaging.internetofthings.ibmcloud.com:1883";
```
Comment (marc2905) [inline bluemix-quickstart/cpp/main.cpp]: ibm?

File: azure-http/cpp/main.cpp [hunk 20]
```diff
@@ -0,0 +1,257 @@
...
+			sprintf_s(msgText, sizeof(msgText),
+					"{\"deviceId\": \"myFirstDevice\",\"temperature\": %.2f}",
+					analogValue);
+			if ((messages[iterator].messageHandle =
+					IoTHubMessage_CreateFromByteArray(
+							(const unsigned char*) msgText, strlen(msgText)))
+					== NULL) {
+				fprintf(stderr, "ERROR: iotHubMessageHandle is NULL!\n");
+			} else {
+
```
Comment (marc2905) [inline azure-http/cpp/main.cpp]: remove empty line

File: azure-http/cpp/main.cpp [hunk 21]
```diff
@@ -0,0 +1,257 @@
...
+	do {
+		if (iterator < MESSAGE_COUNT) {
+			float analogValue = temp_sensor->raw_value();
+			sprintf_s(msgText, sizeof(msgText),
+					"{\"deviceId\": \"myFirstDevice\",\"temperature\": %.2f}",
+					analogValue);
+			if ((messages[iterator].messageHandle =
+					IoTHubMessage_CreateFromByteArray(
+							(const unsigned char*) msgText, strlen(msgText)))
+					== NULL) {
```
Comment (marc2905) [inline azure-http/cpp/main.cpp]: put in one line

File: azure-http/cpp/main.cpp [hunk 22]
```diff
@@ -0,0 +1,257 @@
...
+	if (IoTHubClient_LL_SetOption(iotHubClientHandle, "MinimumPollingTime",
+			&minimumPollingTime) != IOTHUB_CLIENT_OK) {
+		fprintf(stderr, "WARN: FAILED to set option \"MinimumPollingTime\"\n");
+	}
+
+	/* Setting Message call back, so we can receive Commands. */
+	if (IoTHubClient_LL_SetMessageCallback(iotHubClientHandle,
+			ReceiveMessageCallback, &receiveContext) != IOTHUB_CLIENT_OK) {
+		fprintf(stderr,
+				"ERROR: IoTHubClient_LL_SetMessageCallback..........FAILED!\n");
```
Comment (marc2905) [inline azure-http/cpp/main.cpp]: put in one line

File: bluemix-quickstart/cpp/main.cpp (hunk 9 above)
Comment (ThomasLyetPro) [inline bluemix-quickstart/cpp/main.cpp]: Done

File: bluemix-quickstart/cpp/main.cpp (hunk 1 above)
Comment (ThomasLyetPro) [inline bluemix-quickstart/cpp/main.cpp]: Done

File: azure-http/cpp/main.cpp [hunk 23]
```diff
@@ -0,0 +1,259 @@
...
+ * This project template allow you to test the Microsoft* Azure* IoT Hub.
+ * It will read the temperature from the Grove Temperature Sensor and send
+ * it to the Azure* IoT Hub.
+ *
+ * First steps:
+ * - Create an IoT HuB following the instruction in
+ * https://github.com/Azure/azure-iot-sdks/blob/master/doc/setup_iothub.md.
+ * - Create a connection string for a device following the instruction in
+ * https://github.com/Azure/azure-iot-sdks/blob/master/doc/manage_iot_hub.md.
+ * - Copy the connection string inside the "credentials.h" header
```
Comment (marc2905) [inline azure-http/cpp/main.cpp]: please mov description above

**Comment by tingleby** [comment]
Whats the status of this, Is it still draft?

**Comment by ThomasLyetPro** [comment]
Yes, I am waiting for the legal header approval before pushing the final version.

**Comment by ThomasLyetPro** [comment]
Patch updated - contains now example for the 3 protocols: AMQP, MQTT and HTPP

**Comment by tingleby** [comment]
Done
